modules/solid_mechanics/test/neml2/src/Multilinear6DInterpolationModel.cxx

```diff
@@ -0,0 +1,172 @@
+  : Model(options),
+    _s(declare_input_variable<Scalar>("von_mises_stress")),
+    _T(declare_input_variable<Scalar>("temperature")),
+    _ep(declare_input_variable<Scalar>("equivalent_plastic_strain")),
+    _cell_dd(declare_input_variable<Scalar>("cell_dd_density")),
+    _wall_dd(declare_input_variable<Scalar>("wall_dd_density")),
+    _env_fac(declare_input_variable<Scalar>("env_factor")),
+    _output_rate(declare_output_variable<Scalar>("output_rate"))
+{
+  std::string filename = options.get<std::string>("model_file_name");
+  std::ifstream model_file(filename.c_str());
+  model_file >> _json;
+  // storing grid points for indexing.
+  // these should be stored differently so that they are all read in at once.  The order of this can
+  // get messed up easily
+  _stress_grid = json_vector_to_torch("in_stress");
+  _temperature_grid = json_vector_to_torch("in_temperature");
+  _plastic_strain_grid = json_vector_to_torch("in_plastic_strain");
+  _cell_grid = json_vector_to_torch("in_cell");
+  _wall_grid = json_vector_to_torch("in_wall");
+  _env_grid = json_vector_to_torch("in_env");
+
+  // Storing values for interpolation
+  std::string filename_variable = options.get<std::string>("model_file_variable_name");
+  _grid_values = json_6Dvector_to_torch(filename_variable);
+}
+
+void
+Multilinear6DInterpolationModel::set_value(bool /*out*/, bool /*dout_din*/, bool /*d2out_din2*/)
+{
+  neml_assert_dbg("Multilinear6DInterpolationModel should never be called, it is only base class "
+                  "with helper functions.");
+}
+
+std::pair<Scalar, Scalar>
+Multilinear6DInterpolationModel::findLeftIndexAndFraction(const Scalar grid,
+                                                          const Scalar interp_points)
+{
+  // idx is for the left grid point.
+  // searchsorted returns the right idx so -1 makes it the left
+  Scalar left_idx = Scalar(torch::searchsorted(grid, interp_points)) -
+                    torch::tensor(1, default_integer_tensor_options());
+  Scalar left_coord = grid.index({left_idx});
+  Scalar right_coord = grid.index({left_idx + torch::tensor(1, default_integer_tensor_options())});
+  Scalar left_fraction = (right_coord - interp_points) / (right_coord - left_coord);
+  return std::pair<Scalar, Scalar>(left_idx, torch::stack({left_fraction, 1 - left_fraction}, -1));
+}
+
+Scalar
+Multilinear6DInterpolationModel::compute_interpolation(
+    const std::vector<std::pair<Scalar, Scalar>> index_and_fraction, const Scalar grid_values)
+{
+  Scalar result = Scalar::zeros_like(_T);
+  for (const auto i : {0, 1})
+    for (const auto j : {0, 1})
+      for (const auto k : {0, 1})
+        for (const auto l : {0, 1})
+          for (const auto m : {0, 1})
+            for (const auto n : {0, 1})
+            {
+              auto vertex_value =
+                  grid_values.index({(index_and_fraction[0].first +
+                                      torch::tensor(i, default_integer_tensor_options())),
```

Comment:
how do I get rid of torch::tensor(i, default_integer_tensor_options()).   If I just use i it treats it like a double